### PR TITLE
[16] [Part 2] [Integrate] As a user, I can answer question in survey detail

### PR DESCRIPTION
**Pull request by manh-t**
https://github.com/manh-t/flutter-survey/issues/16

## What happened 👀

Fill in data from survey detail and pick the corresponding question type based on API

## Insight 📝

Refactor widget tree of `TakeSurveyPage` to be cleaner

## Proof Of Work 📹

https://user-images.githubusercontent.com/60863885/119686702-cd9a7b00-be70-11eb-8205-df1ed57035dc.mp4

### CODE REVIEW

File: lib/pages/takesurvey/widget/answer_selection.dart [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    );
+  }
+
+  void _debounceTextChanged({required Function onDebounce}) {
+    if (_debounce?.isActive ?? false) {
+      _debounce?.cancel();
+    }
+    _debounce = Timer(const Duration(milliseconds: 300), () {
+      onDebounce();
+    });
```
Comment (markgravity) [inline lib/pages/takesurvey/widget/answer_selection.dart:250]: Do we really need to debounce here, cause we don' request any API or heavy task each time the text changes, right? 🤔

File: lib/pages/takesurvey/take_survey_model.dart [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            .firstWhereOrNull((element) => element.id == questionId);
+        questionForm?.answers.clear();
+        questionForm?.answers.addAll(answers);
+      }
+    }
+  }
+
+  bool validateAnswers(Question question, int totalAnswers) {
+    final answers = _getFilledAnswers(question.id);
+    if ((question.displayType != "intro" && question.displayType != "outro") &&
```
[the remark this answers is not in the document]
Comment (manh-t) [reply lib/pages/takesurvey/take_survey_model.dart:68]: @markgravity yes, gonna change to `enum` for `displayType` 😄

File: lib/pages/takesurvey/widget/answer_selection.dart (hunk 1 above)
Comment (manh-t) [reply lib/pages/takesurvey/widget/answer_selection.dart:250]: @markgravity yes, but I want to avoid writing value continuously to `TakeSurveyForm` when we fill text or select value in dropdown list, it's just a practice and nice to have 😄

File: lib/di/provider/dio_provider.dart [hunk 3]
```diff
@@ -45,8 +45,8 @@ class DioProvider {
     }
 
     return dio
-      ..options.connectTimeout = 3000
-      ..options.receiveTimeout = 5000
+      ..options.connectTimeout = 30000
+      ..options.receiveTimeout = 30000
```
Comment (manh-t) [inline lib/di/provider/dio_provider.dart:49]: I increase timeout because my internet connection is slow 🙏

File: lib/data/survey.dart [hunk 4]
```diff
@@ -25,10 +27,10 @@ class Survey extends Equatable {
...
-        thankEmailAbove: this.thankEmailAbove ?? "",
-        thankEmailBelow: this.thankEmailBelow ?? "",
-        coverImageUrl: this.coverImageUrl ?? "",
-      );
+      id: this.id,
+      title: this.title ?? "",
+      description: this.description ?? "",
+      thankEmailAbove: this.thankEmailAbove ?? "",
+      thankEmailBelow: this.thankEmailBelow ?? "",
+      coverImageUrl: this.coverImageUrl ?? "");
```
Comment (sleepylee) [inline lib/data/survey.dart:35]: I think all of these mapping data could have been done from another PR as a [Backend] work...the integration should only happen between the UI and the data interaction imo... it's quite big to review all as of now 😔

Comment (manh-t) [reply lib/data/survey.dart:35]: sorry for this 😭  I will try to separate these parts out of integrate task

Comment (manh-t) [reply lib/data/survey.dart:35]: @sleepylee I decide to separate the task into 2 parts so you guys will easily review it. This is the first part #81 , and I'll put this PR to draft to complete the rest of the work. Please help me review #81 first, thanks 🙏

**Comment by manh-t** [comment]
this task needs to improve a bit to cleaner I think. Let me refactor them a bit then you guys can review easily 🙏

**Comment by manh-t** [comment]
and one bug found. For `star`, `heart` type, I should add only one answer to the answer form instead of adding all previous answers of rating

**Comment by manh-t** [comment]
@sleepylee @markgravity @luongvo The code looks better now, please help me review it 🙇

File: lib/pages/takesurvey/widget/question_selection.dart [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+          Container(
+            margin: EdgeInsets.only(top: 15),
+            child: Text(
+              data.question.text,
+              style: Theme.of(context).textTheme.subtitle1,
+            ),
+          ),
+          Expanded(
+            child: Center(
+              child: AnswerSelection(data),
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/question_selection.dart:63]: same here, how about `SurveyAnswer`? 
-- TakeSurvey
 - >PageView
     - >SurveyQuestions
        - >SurveyAnswers
        - >SurveyAnswers
        - > ...
      - >SurveyQuestions
        - >SurveyAnswers
        - >SurveyAnswers
        - >...

File: lib/pages/takesurvey/widget/question_selection.dart [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                  AppLocalizations.of(context)!.actionCancel,
+                  style: Theme.of(context)
+                      .textTheme
+                      .bodyText1!
+                      .copyWith(color: AppColors.blackRussian),
+                ),
+              ),
+            ],
+          );
+        }).then((value) {
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/question_selection.dart:121]: - can we make a better name here? Suggesting `shouldExitSurvey`
- perhaps if you make a type check you won't need the below `if (shouldExitSurvey == true)` but just `if (shouldExitSurvey)`?

File: lib/pages/takesurvey/widget/question_selection.dart [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+          onTap: () {
+            _validateThenGoToNextPage(
+                context, question, totalAnswers, totalQuestions);
+          },
+        ),
+      ),
+    );
+  }
+
+  void _validateThenGoToNextPage(BuildContext context, Question question,
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/question_selection.dart:168]: I'm a little bit concern about this as it has quite a lot of logic inside the View to process data before moving forward 🤔 
For example, can we modify the output of the processor inside the `takeSurveyModel` to return something like `UiModel` or `Result` that we can process via a
```
  switch(case):
    Success: {}
    Failed: {}
```
?

Also should the `takeSurveyModel.increasePage()` can be managed inside the `takeSurveyModel`? 🤔

File: lib/pages/takesurvey/widget/answer_selection.dart [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    );
+  }
+
+  Widget _buildSmileyRating(int itemCount, Function onRate) {
+    return RatingBar.builder(
+      itemCount: itemCount,
+      unratedColor: Colors.amber.withOpacity(0.4),
+      direction: Axis.horizontal,
+      itemPadding: EdgeInsets.symmetric(horizontal: 5.0),
+      itemBuilder: (context, _) => Text("😀"),
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/answer_selection.dart:188]: 🤣  no, it has to be 😡 😕😐🙂😄 like the design 🤣  you need to iterate from the `index`

Comment (sleepylee) [reply lib/pages/takesurvey/widget/answer_selection.dart:188]: and remember that it's a single choice i.e: choose 1 (active) and the rest has to be disable (inactive) 🤣

File: lib/pages/takesurvey/widget/answer_selection.dart [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+          widget.data.answers.length,
+          (rating) {
+            _addAnswersForRating(rating);
+          },
+        );
+        break;
+      case DisplayType.smiley:
+        answer = _buildSmileyRating(
+          widget.data.answers.length,
+          (rating) {
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/answer_selection.dart:62]: a bit strange that it breaks the line here 😛

File: lib/pages/takesurvey/widget/answer_selection.dart [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+          Assets.icStarActive,
+          Assets.icStarInactive,
+          widget.data.answers.length,
+          (rating) {
+            _addAnswersForRating(rating);
+          },
+        );
+        break;
+      case DisplayType.heart:
+        answer = _buildRating(
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/answer_selection.dart:50]: I think it's better that we use the named constructor for function returns a widget that has >=3 params like this, it's easier to read and understand the params we pass in. Sorry I missed this from the last UI PR

Comment (sleepylee) [reply lib/pages/takesurvey/widget/answer_selection.dart:50]:
```
_buildRating(
  activeIcon: Assets.icHeartActive,
  inactiveIcon: Assets.icHeartInactive,
  size: widget.data.answers.length,
  onRated: (rating) =>  _addAnswersForRating(rating);
```

File: lib/pages/takesurvey/widget/answer_selection.dart [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+          (rating) {
+            _addAnswersForRating(rating);
+          },
+        );
+        break;
+      case DisplayType.smiley:
+        answer = _buildSmileyRating(
+          widget.data.answers.length,
+          (rating) {
+            _addAnswersForRating(rating);
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/answer_selection.dart:63]: can be shortened as `(rating) => _addAnswersForRating(rating);` I think?

File: lib/pages/takesurvey/take_survey_model.dart [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+            ?.answers ??
+        [];
+  }
+
+  bool _checkFillAllTextFields(String questionId, int totalAnswers) {
+    final questionForm = _takeSurveyForm.questions
+        .firstWhereOrNull((element) => element.id == questionId);
+    return questionForm?.answers.length == totalAnswers &&
+        !(questionForm?.answers.any((element) => element.answer.isEmpty) ==
+            true);
```
Comment (sleepylee) [inline lib/pages/takesurvey/take_survey_model.dart:153]: I'm just a bit curious if we need the `== true` here 😅  didn't run it on the IDE yet so I'm not sure how type is inferred here.

**Review by sleepylee — COMMENTED** [review]
🏃 ....💣

File: lib/pages/takesurvey/widget/take_survey.dart [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            final surveyState = watch(takeSurveyModelProvider);
+            return surveyState.maybeWhen(
+              init: () => Container(
+                child: LoadingIndicator(),
+              ),
+              dataLoaded: (data) => PageView.builder(
+                physics: NeverScrollableScrollPhysics(),
+                controller: pageController,
+                itemCount: data.length,
+                itemBuilder: (context, i) => QuestionSelection(
```
[the remark this answers is not in the document]
Comment (manh-t) [reply lib/pages/takesurvey/widget/take_survey.dart:41]: agree, let me update it 🙏

File: lib/pages/takesurvey/widget/question_selection.dart (hunk 6 above)
Comment (manh-t) [reply lib/pages/takesurvey/widget/question_selection.dart:121]: @sleepylee the `value` object is nullable, so I cannot simply use `if (shouldExitSurvey)` 😞 . I'm going to create a dialog based on the platform in later PR and handle the object. Then we can expose the value as non-nullable 🙏  
anw, let me update the naming of value first 😃

File: lib/pages/takesurvey/widget/question_selection.dart (hunk 7 above)
Comment (manh-t) [reply lib/pages/takesurvey/widget/question_selection.dart:168]: ah, let me handle this part in `takeSurveyModel`, I missed it when refactoring 😓

File: lib/pages/takesurvey/widget/answer_selection.dart (hunk 8 above)
Comment (manh-t) [reply lib/pages/takesurvey/widget/answer_selection.dart:188]: 😐  let me seeeee, if I have to create a new rating widget to fulfill the requirement, I will create a separate PR to handle it 🙏

File: lib/pages/takesurvey/take_survey_model.dart (hunk 12 above)
Comment (manh-t) [reply lib/pages/takesurvey/take_survey_model.dart:153]: let me add the check `questionForm != null` to avoid these conditions like this 😖

File: lib/pages/takesurvey/widget/answer_selection.dart (hunk 8 above)
Comment (manh-t) [reply lib/pages/takesurvey/widget/answer_selection.dart]: @sleepylee the smiley rating widget has been implemented in #84 🙋

**Comment by manh-t** [comment]
@sleepylee Updated 🙌

File: lib/pages/takesurvey/take_survey_model.dart [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  List<AnswerForm> _getFilledAnswerForms(String questionId) {
+    return _takeSurveyForm.questions
+            .firstWhereOrNull((element) => element.id == questionId)
+            ?.answers ??
+        [];
+  }
+
+  bool _checkFillAllTextFields(String questionId, int totalAnswers) {
```
Comment (sleepylee) [inline lib/pages/takesurvey/take_survey_model.dart:157]: maybe `_isAllFieldsNotEmpty` could be a better name, especially when you return `bool`

File: lib/pages/takesurvey/widget/survey_answer.dart [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+            context, widget.data.answers.map((e) => e.text).toList(),
+            onSelect: (index) => takeSurveyModel.addAnswerForDropdown(
+                widget.data.question.id, widget.data.answers[index]));
+        break;
+      default:
+        answer = SizedBox.shrink();
+        break;
+    }
+    return answer;
+  }
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/survey_answer.dart:118]: the whole switch is too big, perhaps you would want to create smaller functions

**Review by sleepylee — COMMENTED** [review]
some last comments mostly on the code style, the integration looks fine to me

**Comment by manh-t** [comment]
Thank @sleepylee for the suggestions, I updated them all 😄

File: lib/pages/takesurvey/take_survey_model.dart [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+        .firstWhereOrNull((element) => element.id == questionId)
+        ?.answers
+        .firstWhereOrNull((element) => element.id == answerId);
+  }
+
+  List<AnswerForm> _getFilledAnswerForms(String questionId) {
+    return _takeSurveyForm.questions
+            .firstWhereOrNull((element) => element.id == questionId)
+            ?.answers ??
+        [];
```
Comment (sleepylee) [inline lib/pages/takesurvey/take_survey_model.dart:153]: just curious, is this because of the auto-format, or can we put it in the above line? `?.answers ?? []`

File: lib/pages/takesurvey/widget/survey_question.dart [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+              child: Icon(
+                Icons.close,
+                color: Colors.white,
+              ),
+              shape: CircleBorder(),
+            ),
+          ),
+          SizedBox(height: 30),
+          Text(
+            "$page/$total",
```
Comment (sleepylee) [inline lib/pages/takesurvey/widget/survey_question.dart:48]: actually, you can leave this logic to the view model (format it and pull it out from `.data`), and reduce the params `page` + `total` from the constructor. But it's fine for now.

**Review by sleepylee — APPROVED** [review]

File: lib/pages/takesurvey/take_survey_model.dart (hunk 16 above)
Comment (manh-t) [reply lib/pages/takesurvey/take_survey_model.dart:153]: yes, it is auto-format 😭  I will try to re-config the formatting or Dart. Currently, it always breaks the line after the operator (`&&`, `||`, `??`) and it looks stupid 🤕

**Review by markgravity — APPROVED** [review]
LGTM 🤖
